### PR TITLE
[clang-tidy] Fix for cppcoreguidelines-pro-type-union-access if memLoc is invalid

### LINKED ISSUE #102945: clang-tidy: [cppcoreguidelines-pro-type-union-access] sometimes does not report location
[cppcoreguidelines-pro-type-union-access] gets triggered on any union member access. It uses `getMemberLoc()` function to retrieve the location of the union member (i.e. `unionMemberA` in `auto i = Z.unionMemberA`) 
here — https://github.com/llvm/llvm-project/blob/11aa31f595325d6b2dede3364e4b86d78fffe635/clang-tools-extra/clang-tidy/cppcoreguidelines/ProTypeUnionAccessCheck.cpp#L26

Sometimes this location is unavailable. For example

```c++
union Z {
  int a;

  template <class TP> operator TP *() const;
};

Z foo();

int main() {
    void *ret = foo(); // Triggers warning
    return 0;
}
```

This results in the confusing message, _note the absence_ of `file:line:col` triple:

```console
warning: do not access members of unions; use (boost::)variant instead [cppcoreguidelines-pro-type-union-access]
```

_____________________

There are two options to fix the issue:

1. Disable report if the member location is invalid:

```c++
 if (auto MemberLoc = Matched->getMemberLoc(); MemberLoc.isValid())
     diag(MemberLoc,
          "do not access members of unions; use (boost::)variant instead");
```

2. Use `Matched` location instead:

```c++
 if (auto MemberLoc = Matched->getMemberLoc(); MemberLoc.isValid())
     diag(MemberLoc,
          "do not access members of unions; use (boost::)variant instead");
 else
     diag(Matched->getExprLoc(),
          "do not access members of unions; use (boost::)variant instead");
```

or similar code.

Please confirm and advise.

### PR DESCRIPTION
**Pull request by ksromanov**
[cppcoreguidelines-pro-type-union-access] clang-tidy check issues a warning, which location points to the union field access using `getMemberLoc()`. However, sometimes `getMemberLoc()` returns invalid location, for instance when the accessed union member is not a data field, but a function. 

The main problem is that the check issues warning without location information is only confusing people, when the checked file contains lots of includes. This situation often occurs in practice.

This PR addresses this situation, returning the original expression location if member location is not valid.

Fixes #102945

### CODE REVIEW

File: clang-tools-extra/clang-tidy/cppcoreguidelines/ProTypeUnionAccessCheck.cpp [hunk 1]
```diff
@@ -23,8 +23,12 @@ void ProTypeUnionAccessCheck::registerMatchers(MatchFinder *Finder) {
...
 void ProTypeUnionAccessCheck::check(const MatchFinder::MatchResult &Result) {
   const auto *Matched = Result.Nodes.getNodeAs<MemberExpr>("expr");
-  diag(Matched->getMemberLoc(),
-       "do not access members of unions; use (boost::)variant instead");
+  if (auto MemberLoc = Matched->getMemberLoc(); MemberLoc.isValid())
+    diag(MemberLoc,
+         "do not access members of unions; use (boost::)variant instead");
+  else
+    diag(Matched->getBeginLoc(),
+         "do not access members of unions; use (boost::)variant instead");
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/cppcoreguidelines/ProTypeUnionAccessCheck.cpp:31]:
```suggestion
  SourceLocation Loc = Matched->getMemberLoc();
  if (Loc.isInvalid())
     Loc = Matched->getBeginLoc();
  diag(Loc, "do not access members of unions; use (boost::)variant instead");
```

re-format may be needed

**Review by PiotrZSL — CHANGES_REQUESTED** [review]
Missing release notes entry, about what's fixed in check.
For me this is an full-fix, not an workaround.

**Review by 5chmidti — CHANGES_REQUESTED** [review]
This is still missing a release note, see https://github.com/llvm/llvm-project/blob/40975da950c95124155b752cd683d945f7d203fd/clang-tools-extra/docs/ReleaseNotes.rst?plain=1#L104 for where to add it and how they look like

Functionality-wise, this looks good

**Comment by ksromanov** [comment]
> This is still missing a release note, see

Many thanks! This is my first PR to `llvm`, I didn't know about it, sorry.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -116,6 +116,10 @@ Changes in existing checks
   <clang-tidy/checks/readability/redundant-smartptr-get>` check to
   remove `->`, when redundant `get()` is removed.
 
+- Fixed :doc:`cppcoreguidelines-pro-type-union-access
+  <clang-tidy/checks/cppcoreguidelines/pro-type-union-access>` check to
+  report location even when member location is not valid.
```
Comment (5chmidti) [inline clang-tools-extra/docs/ReleaseNotes.rst:121]: nit: `... report a location even when the member ...`

**Review by 5chmidti — APPROVED** [review]

File: clang-tools-extra/docs/ReleaseNotes.rst (hunk 2 above)
Comment (ksromanov) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Thanks! Articles will always be a problem for me. :-(

**Comment by ksromanov** [comment]
@PiotrZSL would you please review once again?

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -141,6 +141,10 @@ New check aliases
 Changes in existing checks
 ^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+- Fixed :doc:`cppcoreguidelines-pro-type-union-access
+  <clang-tidy/checks/cppcoreguidelines/pro-type-union-access>` check to
+  report a location even when the member location is not valid.
```
Comment (nicovank) [inline clang-tools-extra/docs/ReleaseNotes.rst:146]: Order alphabetically by check name.

**Review by nicovank — APPROVED** [review]
LGTM! Minus alphabetical ordering in release notes. I'll merge this tomorrow to give time to others just in case.

FWIW this issue actually triggers a crash/assertion failure when those are enabled:
```
clang-tidy: clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.cpp:178: DiagnosticBuilder clang::tidy::ClangTidyContext::diag(StringRef, SourceLocation, StringRef, DiagnosticIDs::Level): Assertion `Loc.isValid()' failed.
[...]
```

File: clang-tools-extra/docs/ReleaseNotes.rst (hunk 3 above)
Comment (ksromanov) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Done.

**Comment by nicovank** [comment]
There's a CI failure but looks unrelated. If you have a minute please rebase/merge main to trigger the checks again and hopefully the problem disappears. We can also just keep an eye on the buildbots post-merge tomorrow.

**Comment by ksromanov** [comment]
> EDIT: Turns out I can just do it, never knew I could press that button on the PR. Anyway should re-trigger things :).

Thank you!

**Comment by ksromanov** [comment]
Thanks a lot!

**Pull request merged**